danascheider/skyrim_inventory_management_frontend
# Create DashboardContext

## Context

[**Create DashboardContext**](https://trello.com/c/qY36NlqF/27-create-dashboardcontext)

There are a number of components that rely on dashboard data now. I wanted to create a React context that would enable me to keep this data in one place across components and move as many API calls as possible to `Page` components, allowing child components to consume this data while keeping thee responsibility for fetching it in one place.

## Changes

* Create DashboardContext that provides `token`, `profileData`, `setSessionCookie`, `removeSessionCookie`, and `setProfileData` properties
* Use `DashboardContextProvider` to wrap `LoginPage`, `DashboardPage`, and `ShoppingListPage` components
* Consume the context from the wrapped pages and the `DashboardHeader` component
* Move responsibility for fetching profile data and redirecting on a 401 to the `DashboardProvider` (login belongs to the `LoginPage`)
* Add PR template

## Considerations

I tried a number of approaches and the one I settled on was not my first choice but it was the way that reliably worked. I spent a day experimenting with an `AuthContext` that would provide login and logout functions and it just didn't work out. There were issues with `gapi`, which doesn't have an npm package or play well with React. It worked better to put code handling login and logout in or close to the components that actually use those functions. Unfortunately, that means the logout function still lives in the `DashboardHeader`.

I also had to use kind of a hack to get Storybook working. I experimented with the [storybook-react-context](https://storybook.js.org/addons/storybook-react-context/) addon but couldn't get it to work or even change anything at all about how Storybook was working. So I added an `overrideValue` prop to the `DashboardProvider` enabling the provider's value to be manipulated. I really don't like putting test wiring in the app code, but writing my own addon or decorator was something I didn't have the time or patience to learn how to do right now.

## Manual Test Steps

This is a refactor, so it should not change anything about how the app actually works, and does not do so in my dev environment or Storybook (other than the hack I described that was required to kind-of-mock the provider data). Make sure you can log in, log out from the shopping list or dashboard page, and navigate between pages without getting logged out.
